src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,295 @@
+      || !os_xlib_dlsym("XCloseDisplay", (void **)&_XCloseDisplay, err)
+      || !os_xlib_dlsym("XkbGetState", (void **)&_XkbGetState, err)) {
+    return false;
+  }
+  return xlib_opened;
+}
+
+/// Get information about Num/Caps/Scroll Lock state in Linux
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a pointer to enum KbdLocks.
+/// @param[out]  err  Location and description where error message is to
+///                   be saved.
+///
+/// @return -1 in case of error otherwise an integer specified for given OS.
+static mod_mask_t os_get_locks_status(mod_mask_t *const mods, Error *const err)
+{
+  if (!os_xlib_init(err)) {
+    return -1;
+  }
+
+  _XDisplay *const display = _XOpenDisplay(NULL);
+  if (display == NULL) {
+    api_set_error(err, kErrorTypeException,
+                  "E5666: Unable to open the display\n"
+                  "XOpenDisplay Failed");
+    return -1;
+  }
+
+  _XkbStateRec xkb_state;
+  const _XStatus status = _XkbGetState(display, _XkbUseCoreKbd, &xkb_state);
+  if (status) {
+    api_set_error(err, kErrorTypeException,
+                  "E5888: Unable to get keyboard state of the display\n"
+                  "XkbGetState Failed: %d", status);
+
+    _XCloseDisplay(display);
+  }
+
+  int state = xkb_state.data[kLockedModsOffset];
+  if (state & xNumLock) {
+    *mods |= kNumLock;
+  }
+
+  if (state & xCapsLock) {
+    *mods |= kCapsLock;
+  }
+
+  if (state & xScrollLock) {
+    *mods |= xScrollLock;
+  }
+
+  _XCloseDisplay(display);
+  return kCapsLock | kNumLock | kScrollLock;
+}
+
+#elif defined(WIN32)
+/// Get information about Num/Caps/Scroll Lock state in Windows.
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a pointer to enum KbdLocks.
+/// @param[out]  err  Location and description where error message is to
+///                   be saved.
+///
+/// @return -1 in case of error otherwise an integer specified for given OS.
+static mod_mask_t os_get_locks_status(mod_mask_t *const mods, Error *const err)
+{
+  *mods = 0;
+  if (GetKeyState(VK_CAPITAL) & 0x0001) {
+    *mods |= kCapsLock;
+  }
+  if (GetKeyState(VK_NUMLOCK) & 0x0001) {
+    *mods |= kNumLock;
+  }
+  if (GetKeyState(VK_SCROLL) & 0x0001) {
+    *mods |= kScrollLock;
+  }
+  return kCapsLock | kNumLock | kScrollLock;
+}
+
+#elif defined(__APPLE__)
+/// Get information about Num/Caps Lock state in MacOS
+///
+/// To be used in os_mods_status() function.
+///
+/// @param[out]  mods  Holds the information about the status of various locks
+///                    in form of a pointer to enum KbdLocks.
+/// @param[out]  err  Location and description where error message is to
+///                   be saved.
+///
+/// @return -1 in case of error otherwise an integer specified for given OS.
```

Comment:
This is ambiguous: what is “an intereg specified for given OS” supposed to mean? Should be something like “-1 in case of error and a mask specifying which values in `mods` are valid otherwise” (no mention of OS because it is really not OS-specific, just we do not support linux terminal or wayland currently and have no providers either).

Reply:
OK, I thought the OS or platform determines it, like I learn that scroll lock was in extended keyboard of MacOS, so thought about it. But yes, its a mask, better to write the mask